frontend/routes/modelSettings/modelSettings.component.tsx

```diff
@@ -0,0 +1,390 @@
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import * as React from 'react';
+import * as queryString from 'query-string';
+import { Formik, Field } from 'formik';
+import { snakeCase } from 'lodash';
+
+import Button from '@material-ui/core/Button';
+import Grid from '@material-ui/core/Grid';
+import InputLabel from '@material-ui/core/InputLabel';
+import Input from '@material-ui/core/Input';
+
+import { clientConfigService } from '../../services/clientConfig';
+
+import { Panel } from '../components/panel/panel.component';
+import { CellSelect } from '../components/customTable/components/cellSelect/cellSelect.component';
+
+import {
+	FieldsRow,
+	StyledTextField,
+	SelectWrapper,
+	StyledChip,
+	TopicTypesContainer,
+	StyledForm,
+	Headline,
+	TypesGrid,
+	GridColumn
+} from './modelSettings.styles';
+
+const PANEL_PROPS = {
+	title: 'Model Settings',
+	paperProps: {
+		height: '100%'
+	}
+};
+
+const ENTER_KEY = 'Enter';
+
+export const unitsMap = {
+	ft: "Feet and inches",
+	mm: "Millimetres",
+	cm: "Centimetres",
+	dm: "Decimetres",
+	m: "Metres"
+};
+
+interface IState {
+	topicTypes?: any[];
+	latitude?: number;
+	longitude?: number;
+	axisX?: number;
+	axisY?: number;
+	axisZ?: number;
+	elevation?: number;
+	angleFromNorth?: number;
+	code?: string;
+	unit?: string;
+}
+
+interface IProps {
+	location: any;
+	fetchModelSettings: (teamspace, modelId) => void;
+	updateModelSettings: (teamspace, modelId, settings) => void;
+	modelSettings: any;
+	currentTeamspace: string;
+}
+
+export class ModelSettings extends React.PureComponent<IProps, IState> {
+	public state = {
+		topicTypes: [],
+		latitude: 0,
+		longitude: 0,
+		axisX: 0,
+		axisY: 0,
+		axisZ: 0,
+		elevation: 0,
+		angleFromNorth: 0
+	};
+
+	public componentDidMount() {
+		const queryParams = queryString.parse(this.props.location.search);
+		const { modelId, targetAcct } = queryParams; // TODO: change targetAcct name
+
+		this.props.fetchModelSettings(targetAcct, modelId);
+	}
+
+	public componentDidUpdate(prevProps, prevState) {
+		const properties = this.props.modelSettings.properties;
+		const topicTypes = properties.topicTypes;
+		const surveyPoints = this.props.modelSettings.surveyPoints;
+		const prevSurveyPoints = prevProps.modelSettings.surveyPoints;
+		const elevation = this.props.modelSettings.elevation;
+		const angleFromNorth = this.props.modelSettings.angleFromNorth;
+
+		if (!prevProps.modelSettings.properties && properties) {
+			if (topicTypes && topicTypes.length !== prevState.topicTypes.length) {
+				this.setState({ topicTypes });
```

Comment:
Each `.setState` invokes async state change. I think it's better to collect these changes and then call `.setState` once.

```suggestion
const changes = {};

if(something) {
   changes.foo = 'bar';
}

if (!isEmpty(changes)){
   this.setState(changes);
}
```